hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/SerialConsumer.java

```diff
@@ -32,139 +28,120 @@
         this.trackers = trackers;
         this.messageConverterResolver = messageConverterResolver;
-        this.topic = topic;
         this.messageReceiver = () -> {
             throw new IllegalStateException("Consumer not initialized");
         };
+        this.topic = topic;
     }
 
-    private int getMaxPermits(Subscription subscription) {
-        return ofNullable(subscription.getSerialSubscriptionPolicy().getInflightSize())
-                .map(value -> Math.min(value, defaultInflight))
-                .orElse(defaultInflight);
+    private int calculateInflightSize(Subscription subscription) {
+        return Math.min(
+                subscription.getSerialSubscriptionPolicy().getInflightSize(),
+                defaultInflight
+        );
     }
 
     private String getId() {
         return subscription.getId();
     }
 
     @Override
-    public void run() {
-        try {
-            setThreadName();
-
-            logger.info("Starting consumer for subscription {} ", subscription.getId());
-
-            Timer.Context timer = new Timer().time();
-            this.messageReceiver = initializeMessageReceiver();
-            rateLimiter.initialize();
-
-            logger.info("Started consumer for subscription {} in {} ms", subscription.getId(), TimeUnit.NANOSECONDS.toMillis(timer.stop()));
-
-            startConsumption(messageReceiver);
-        } finally {
-            unsetThreadName();
-            ofNullable(messageReceiver).ifPresent(MessageReceiver::stop);
-            logger.info("Stopped consumer for subscription {}", subscription.getId());
-            stoppedLatch.countDown();
-        }
-    }
-
-    private void startConsumption(MessageReceiver messageReceiver) {
-        while (isConsuming()) {
-            try {
-                inflightSemaphore.acquire();
-
-                Message message = messageReceiver.next();
-
-                Message convertedMessage = messageConverterResolver.converterFor(message, subscription).convert(message, topic);
-
-                sendMessage(convertedMessage);
-            } catch (MessageReceivingTimeoutException messageReceivingTimeoutException) {
-                inflightSemaphore.release();
-                logger.debug("Timeout while reading message from topic. Trying to read message again", messageReceivingTimeoutException);
-            } catch (Exception e) {
-                logger.error("Consumer loop failed for " + getId(), e);
-            }
-        }
-    }
-
-    private MessageReceiver initializeMessageReceiver() {
+    public void consume(Runnable signalsInterrupt) {
         try {
-            logger.debug("Consumer: preparing message receiver for subscription {}", subscription.getId());
-            return messageReceiverFactory.createMessageReceiver(topic, subscription);
+            do {
+                signalsInterrupt.run();
+            } while (!inflightSemaphore.tryAcquire(500, TimeUnit.MILLISECONDS));
```

Comment:
I think that 500 ms looks like a decent default but should be adjusted based on output rate and desired consumer responsiveness.